pkg/web: restrict handler types

tombrk opened the pull request:
Makes `pkg/web` only accept handles from the following set:

```go
	handlerStd       = func(http.ResponseWriter, *http.Request)
	handlerStdCtx    = func(http.ResponseWriter, *http.Request, *web.Context)
	handlerStdReqCtx = func(http.ResponseWriter, *http.Request, *models.ReqContext)
	handlerReqCtx    = func(*models.ReqContext)
	handlerReqCtxRes = func(*models.ReqContext) Response
	handlerCtx       = func(*web.Context)
```

This is a first step to reducing above set to only `http.Handler`.

---

Due to a cyclic import situation between `pkg/models` and `pkg/web`, parts of this PR were put into `pkg/api/response`, even though they definitely do not belong there. This however is _temporary_ until we untangle `models.ReqContext`.

tombrk committed:
pkg/web: restrict valid handler types

tombrk committed:
pkg/web: HandlerType insights

set HANDLER_SUMMARY=1 to get a summary of routes and their handlertypes
printed

tombrk committed:
pkg/web: handle returned response.Response

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=48495) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=48495) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=48495) it.</sub>

sdboyer commented:
re: these import cycles: i think they're gonna be pretty much inevitable.

If the eventual goal is to define patterns for our overall web framework, and the handlers originate from within componentized services, then i think a registry pattern works well in order to avoid the cycle.

![component registration pattern](https://user-images.githubusercontent.com/21599/165926026-c25ba74e-0a7d-483d-9545-cd9937d0b99d.png)

This is the pattern we followed in #44840. In the context of the above diagram, we:
* Declared shared interfaces for components in `internal/framework/kubecontroller`
* Declared instances of the above shared interfaces + k8s-style schemas in a component pkg (e.g. `internal/coremodel/datasource/crd`)
* Pulled in all the above component instances from a registry package `internal/framework/kubecontroller/staticregistry`
  * (That PR currently contains a little more indirection with `pkg/schema` but that would've eventually gone away)
* Referenced the registry in a serverbuilder package (`internal/k8sbridge`)
* Drew it all together into a new http server package (`internal/intentapi`)

Parts of this pattern are now already merged in the form of coremodels (`pkg/framework/coremodel`, `pkg/framework/coremodel/staticregistry`, `pkg/coremodel/dashboard`), and other parts will land with #48252

tombrk committed:
pkg/web: move hack to pkg/response

Temporarily moves the hacky parts into pkg/response, as it is nearly
always in the dependency graph, so we avoid link issues

tombrk committed:
*: ensure reqCtx is in context in tests

makes sure the reqCtx is always inside the context.Context of
http.Request, so that extracting without the macaron injection system works

tombrk committed:
*: merge main to fix conflicts

tombrk committed:
*: linter fixes

sh0rez commented on pkg/api/common_test.go:
```diff
@@ -411,6 +412,9 @@ func setupHTTPServerWithCfgDb(t *testing.T, useFakeAccessControl, enableAccessCo
 		initCtx.Logger = log.New("api-test")
 		c.Map(initCtx)
+
+		c.Req = c.Req.WithContext(ctxkey.Set(c.Req.Context(), initCtx))
+		c.Map(c.Req)
```
Because tests mess with the internals of `pkg/web`, we need to do this several times here.

we could possibly modify `Mux.Map()` so that it detects `models.ReqContext` above and does the mapping automatically.

I dislike both though. Opinions @zserge @sakjur?

tombrk committed:
*: cleanup

vtorosyan commented:
@sh0rez Just a quick follow up on this change. It unfortunately broke the OAuth sign in, please see this issue for the reference https://github.com/grafana/grafana/pull/49630

Unfortunately the fixes in the #49630 are introducing different problems and we are still searching for a proper fix, so if please let us know if there are any better ideas.

vtorosyan commented:
An alternative proposal - https://github.com/grafana/grafana/pull/49653
